File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/VB/UtilitiesLibrary/D3DUtilities/Texture/TextureLoader.vb
' Copyright (c) Microsoft Corporation.  All rights reserved.


Imports Microsoft.VisualBasic
Imports System
Imports System.IO
Imports System.Runtime.InteropServices


Imports Microsoft.WindowsAPICodePack.DirectX.Direct3D10
Imports Microsoft.WindowsAPICodePack.DirectX.Graphics
Imports Microsoft.WindowsAPICodePack.DirectX.WindowsImagingComponent

Namespace Microsoft.WindowsAPICodePack.DirectX.DirectXUtilities
	Public Class TextureLoader
		''' <summary>
		''' Creates a ShaderResourceView from a bitmap in a Stream. 
		''' </summary>
		''' <param name="device">The Direct3D device that will own the ShaderResourceView</param>
		''' <param name="stream">Any Windows Imaging Component decodable image</param>
		''' <returns></returns>
		Public Shared Function LoadTexture(ByVal device As D3DDevice, ByVal stream As Stream) As ShaderResourceView
            Dim factory As ImagingFactory

            factory = ImagingFactory.Create()

            Dim bitmapDecoder As BitmapDecoder = factory.CreateDecoderFromStream(stream, DecodeMetadataCacheOption.OnDemand)

			If bitmapDecoder.FrameCount = 0 Then
				Throw New ArgumentException("Image file successfully loaded, but it has no image frames.")
			End If

			Dim bitmapFrameDecode As BitmapFrameDecode = bitmapDecoder.GetFrame(0)
			Dim bitmapSource As BitmapSource = bitmapFrameDecode.ToBitmapSource()

			' create texture description
            Dim textureDescription As New Texture2DDescription() With _
            { _
                .Width = bitmapSource.Size.Width, _
                .Height = bitmapSource.Size.Height, _
                .MipLevels = 1, _
                .ArraySize = 1, _
                .Format = Format.R8G8B8A8UNorm, _
                .SampleDescription = New SampleDescription() With _
                { _
                    .Count = 1, _
                    .Quality = 0 _
                }, _
                .Usage = Usage.Dynamic, _
                .BindingOptions = BindingOptions.ShaderResource, _
                .CpuAccessOptions = CpuAccessOptions.Write, _
                .MiscellaneousResourceOptions = MiscellaneousResourceOptions.None _
            }

			' create texture
			Dim texture As Texture2D = device.CreateTexture2D(textureDescription)

			' Create a format converter
            Dim converter As FormatConverter = factory.CreateFormatConverter()
            converter.Initialize(bitmapSource, PixelFormats.Prgba32Bpp, BitmapDitherType.None, BitmapPaletteType.Custom)

			' get bitmap data
			Dim buffer() As Byte = converter.CopyPixels()

			' Copy bitmap data to texture
            Dim texmap As MappedTexture2D = texture.Map(0, Map.WriteDiscard, Direct3D10.MapOptions.None)
			Marshal.Copy(buffer, 0, texmap.Data, buffer.Length)
			texture.Unmap(0)

			' create shader resource view description
            Dim srvDescription As New ShaderResourceViewDescription() With _
            { _
                .Format = textureDescription.Format, _
                .ViewDimension = ShaderResourceViewDimension.Texture2D, _
                .Texture2D = New Texture2DShaderResourceView() With _
                { _
                    .MipLevels = textureDescription.MipLevels, _
                    .MostDetailedMip = 0 _
                } _
            }

			' create shader resource view from texture
			Return device.CreateShaderResourceView(texture, srvDescription)
		End Function
	End Class
End Namespace

File: main/contrib/Windows API Code Pack 1.1/source/Samples/DirectX/CS/UtilitiesLibrary/D3DUtilities/Texture/TextureLoader.cs
// Copyright (c) Microsoft Corporation.  All rights reserved.

using System;
using System.IO;
using System.Runtime.InteropServices;


using Microsoft.WindowsAPICodePack.DirectX.Direct3D10;
using Microsoft.WindowsAPICodePack.DirectX.Graphics;
using Microsoft.WindowsAPICodePack.DirectX.WindowsImagingComponent;

namespace Microsoft.WindowsAPICodePack.DirectX.DirectXUtilities
{
    public class TextureLoader
    {
        /// <summary>
        /// Creates a ShaderResourceView from a bitmap in a Stream. 
        /// </summary>
        /// <param name="device">The Direct3D device that will own the ShaderResourceView</param>
        /// <param name="stream">Any Windows Imaging Component decodable image</param>
        /// <returns>A ShaderResourceView object with the loaded texture.</returns>
        public static ShaderResourceView LoadTexture( D3DDevice device, Stream stream )
        {
            ImagingFactory factory = ImagingFactory.Create();

            BitmapDecoder bitmapDecoder =
                factory.CreateDecoderFromStream(
                    stream,
                    DecodeMetadataCacheOption.OnDemand );

            return LoadFromDecoder(device, factory, bitmapDecoder);
        }

        /// <summary>
        /// Creates a ShaderResourceView from a bitmap in a file. 
        /// </summary>
        /// <param name="device">The Direct3D device that will own the ShaderResourceView</param>
        /// <param name="filePath">Any Windows Imaging Component decodable image</param>
        /// <returns>A ShaderResourceView object with the loaded texture.</returns>
        public static ShaderResourceView LoadTexture(D3DDevice device, String filePath)
        {
            ImagingFactory factory = ImagingFactory.Create();

            BitmapDecoder bitmapDecoder =
                factory.CreateDecoderFromFileName(
                    filePath,
                    DesiredAccess.Read,
                    DecodeMetadataCacheOption.OnDemand);

            return LoadFromDecoder(device, factory, bitmapDecoder);
        }
        private static ShaderResourceView LoadFromDecoder(D3DDevice device, ImagingFactory factory, BitmapDecoder bitmapDecoder)
        {
            if (bitmapDecoder.FrameCount == 0)
                throw new ArgumentException("Image file successfully loaded, but it has no image frames.");

            BitmapFrameDecode bitmapFrameDecode = bitmapDecoder.GetFrame(0);
            BitmapSource bitmapSource = bitmapFrameDecode.ToBitmapSource();

            // create texture description
            Texture2DDescription textureDescription = new Texture2DDescription()
            {
                Width = bitmapSource.Size.Width,
                Height = bitmapSource.Size.Height,
                MipLevels = 1,
                ArraySize = 1,
                Format = Format.R8G8B8A8UNorm,
                SampleDescription = new SampleDescription()
                {
                    Count = 1,
                    Quality = 0,
                },
                Usage = Usage.Dynamic,
                BindingOptions = BindingOptions.ShaderResource,
                CpuAccessOptions = CpuAccessOptions.Write,
                MiscellaneousResourceOptions = MiscellaneousResourceOptions.None
            };

            // create texture
            Texture2D texture = device.CreateTexture2D(textureDescription);

            // Create a format converter
            FormatConverter converter = factory.CreateFormatConverter();
            converter.Initialize(
                bitmapSource,
                PixelFormats.Prgba32Bpp,
                BitmapDitherType.None,
                BitmapPaletteType.Custom);

            // get bitmap data
            byte[] buffer = converter.CopyPixels();

            // Copy bitmap data to texture
            MappedTexture2D texmap = texture.Map(0, Map.WriteDiscard, Microsoft.WindowsAPICodePack.DirectX.Direct3D10.MapOptions.None);
            Marshal.Copy(buffer, 0, texmap.Data, buffer.Length);
            texture.Unmap(0);

            // create shader resource view description
            ShaderResourceViewDescription srvDescription = new ShaderResourceViewDescription()
            {
                Format = textureDescription.Format,
                ViewDimension = ShaderResourceViewDimension.Texture2D,
                Texture2D = new Texture2DShaderResourceView()
                {
                    MipLevels = textureDescription.MipLevels,
                    MostDetailedMip = 0
                }
            };

            // create shader resource view from texture
            return device.CreateShaderResourceView(texture, srvDescription);
        }
    }
}
